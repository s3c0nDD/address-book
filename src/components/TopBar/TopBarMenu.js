import React, { useMemo } from 'react';
import { Link } from 'react-router-dom';

import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';

import { ROUTES } from '../../Routes';

const TopBar = (props) => {
  const { anchorEl, handleClose, isOpen } = props;

  const originPositionObj = useMemo(() => ({
    vertical: 'top',
    horizontal: 'left',
  }), []);

  return (
    <Menu
      anchorEl={anchorEl}
      anchorOrigin={originPositionObj}
      keepMounted
      transformOrigin={originPositionObj}
      open={isOpen}
      onClose={handleClose}
    >
      <MenuItem
        onClick={handleClose}
        component={Link}
        to={ROUTES.root}
      >
        Users Gallery
      </MenuItem>
      <MenuItem
        onClick={handleClose}
        component={Link}
        to={ROUTES.settings}
      >
        Settings
      </MenuItem>
    </Menu>
  )
};

export default TopBar;