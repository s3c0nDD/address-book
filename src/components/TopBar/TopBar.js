import React from 'react';
import PropTypes from 'prop-types';
import { useLocation } from 'react-router-dom'

import AppBar from '@material-ui/core/AppBar';
import Hidden from '@material-ui/core/Hidden';
import IconButton from '@material-ui/core/IconButton';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import MenuIcon from '@material-ui/icons/Menu';

import { ROUTES } from '../../Routes';
import useMenu from '../../hooks/useMenu';
import TopBarMenu from './TopBarMenu';
import TopBarSearch from './TopBarSearch';

/**
 * Top bar component
 * @param {object} props component's props
 * @param {boolean} props.searchDisabled flag is search is disabled
 * @param {string} props.searchText search text state
 * @param {function} props.onSearchChange search state change handler
 */
const TopBar = ({ searchDisabled, searchText, onSearchChange, ...props }) => {
  const { anchorEl, isOpen, handleClose, handleOpen } = useMenu();
  const location = useLocation();

  /**
   * Handle search change
   * @param {object<SyntheticEvent>} event react event
   */
  const handleChange = (event) => {
    onSearchChange(event.target.value);
  };

  const showSearch = (location.pathname === ROUTES.root);

  return (
    <AppBar
      position="fixed"
      {...props}
    >
      <TopBarMenu
        anchorEl={anchorEl}
        handleClose={handleClose}
        open={isOpen}
      />
      <Toolbar>
        <IconButton
          edge="start"
          color="inherit"
          aria-label="menu"
          onClick={handleOpen}
        >
          <MenuIcon />
        </IconButton>
        <Hidden
          only={'xs'}
        >
          <Typography
            variant="h6"
            color="inherit"
          >
            Address Book
          </Typography>
        </Hidden>
        {showSearch && (
          <TopBarSearch
            text={searchText}
            disabled={searchDisabled}
            onTextChange={handleChange}
          />
        )}
      </Toolbar>
    </AppBar>
  )
};

export default TopBar;

TopBar.propTypes = {
  searchDisabled: PropTypes.bool.isRequired,
  searchText: PropTypes.string,
  onSearchChange: PropTypes.func.isRequired,
  props: PropTypes.arrayOf(PropTypes.any)
};