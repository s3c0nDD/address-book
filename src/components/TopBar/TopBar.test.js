import React from 'react';
import routeData from 'react-router';
import { createShallow } from '@material-ui/core/test-utils';

import TopBar from './TopBar';
import TopBarSearch from './TopBarSearch'

describe('TopBar', () => {
  let shallow;
  let component;
  let wrapper;
  let handleChangeMock;

  beforeAll(() => {
    shallow = createShallow();
    handleChangeMock = jest.fn();
    jest.spyOn(routeData, 'useLocation').mockImplementation(() => ({
      pathname: '/', hash: '', search: '', state: ''
    }));
  });

  beforeEach(() => {
    component = (
      <TopBar
        searchDisabled={false}
        searchText="example search"
        onSearchChange={handleChangeMock}
      />
    );
    wrapper = shallow(component);
  });

  it('shallowly renders and matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls change handler on TopBarSearch text change', () => {
    const searchWrapper = wrapper.find(TopBarSearch);
    const mockedEvent = { target: { value: searchWrapper.props().text }};
    searchWrapper.props().onTextChange(mockedEvent);
    expect(handleChangeMock).toBeCalledTimes(1);
    expect(handleChangeMock).toBeCalledWith('example search');
  });

  it('shows TopBarSearch when route is root', () => {
    const searchWrapper = wrapper.find(TopBarSearch);
    expect(searchWrapper).toBeTruthy();
  });
});
