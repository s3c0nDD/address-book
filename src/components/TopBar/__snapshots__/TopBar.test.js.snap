// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar shallowly renders and matches snapshot 1`] = `
<WithStyles(ForwardRef(AppBar))
  position="fixed"
>
  <TopBarMenu
    anchorEl={null}
    handleClose={[Function]}
    open={false}
  />
  <WithStyles(ForwardRef(Toolbar))>
    <WithStyles(ForwardRef(IconButton))
      aria-label="menu"
      color="inherit"
      edge="start"
      onClick={[Function]}
    >
      <MenuIcon />
    </WithStyles(ForwardRef(IconButton))>
    <Hidden
      only="xs"
    >
      <WithStyles(ForwardRef(Typography))
        color="inherit"
        variant="h6"
      >
        Address Book
      </WithStyles(ForwardRef(Typography))>
    </Hidden>
    <TopBarSearch
      disabled={false}
      onTextChange={[Function]}
      text="example search"
    />
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
